desktop/cmp/dash/canvas/DashCanvasModel.js

```diff
@@ -0,0 +1,315 @@
+     * @param {Array} [c.extraMenuItems] - Array of RecordActions, configs or token strings, with
+     *      which to create additional context menu items. Extra menu items will appear
+     *      in the container's context menu below the 'Add' action, and in the 'Options' context
+     *      menus for individual views within the container
+     */
+    constructor({
+        viewSpecs,
+        viewSpecDefaults,
+        initialState = {layout: [], viewState: {}},
+        layoutLocked = false,
+        contentLocked = false,
+        renameLocked = false,
+        persistWith = null,
+        columns = 8,
+        rowHeight = 50,
+        compact = true,
+        margin = [10, 10],
+        maxRows = Infinity,
+        containerPadding = null,
+        extraMenuItems
+    }) {
+        super();
+        makeObservable(this);
+        viewSpecs = viewSpecs.filter(it => !it.omit);
+        ensureUniqueBy(viewSpecs, 'id');
+        this.viewSpecs = viewSpecs.map(cfg => {
+            return new DashCanvasViewSpec(defaultsDeep({}, cfg, viewSpecDefaults));
+        });
+
+        this.restoreState = {initialState, layoutLocked, contentLocked, renameLocked, columns,
+            rowHeight, compact, margin, maxRows, containerPadding};
+        this.layoutLocked = layoutLocked;
+        this.contentLocked = contentLocked;
+        this.renameLocked = renameLocked;
+        this.columns = columns;
+        this.rowHeight = rowHeight;
+        this.maxRows = maxRows;
+        this.containerPadding = containerPadding;
+        this.maxRows = maxRows;
+        this.margin = margin;
+        this.containerPadding = containerPadding;
+        this.compact = compact;
+        this.extraMenuItems = extraMenuItems;
+
+        // Read state from provider -- fail gently
+        let persistState = null;
+        if (persistWith) {
+            try {
+                this.provider = PersistenceProvider.create({path: 'dashCanvas', ...persistWith});
+                persistState = this.provider.read();
+            } catch (e) {
+                console.error(e);
+                XH.safeDestroy(this.provider);
+                this.provider = null;
+            }
+        }
+
+        this.setState(persistState?.state ?? initialState);
+
+        this.addReaction({
+            track: () => [this.state],
+            run: () => this.publishState()
+        });
+    }
+
+    /**
+     * Restore the initial state as specified by the application at construction time. This is the
+     * state without any persisted state or user changes applied.
+     *
+     * This method will clear the persistent state saved for this component, if any.
+     */
+    restoreDefaults() {
+        const {restoreState} = this;
+        this.layoutLocked = restoreState.layoutLocked;
+        this.contentLocked = restoreState.contentLocked;
+        this.renameLocked = restoreState.renameLocked;
+        this.setColumns(restoreState.columns);
```

Comment:
I would just make this method an `@action` and do direct assignments to the bindable properties here for consistency